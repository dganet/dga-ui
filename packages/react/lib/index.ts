import { ButtonComponent } from "./components/Button/ButtonComponent";

export {
  ButtonComponent as Button,
}