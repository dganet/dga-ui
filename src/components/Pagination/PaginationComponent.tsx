import React, { useEffect, useState } from 'react'

import { useTheme } from '@hooks/useTheme'
import { PaginationProps } from './PaginationComponent.types'
import {
    Container,
    PaginationItem,
    ActionButton
} from './PaginationComponent.styles'
import { BsArrowRightShort, BsArrowLeftShort } from 'react-icons/bs'

const PaginationComponent = ({
    count,
    onChange,
    defaultPage = 1
}: PaginationProps) => {
    const theme = useTheme()
    const [currentPage, setCurrentPage] = useState(defaultPage)
    const [pageList, setPageList] = useState([])

    useEffect(() => {
        for (let index = 1; index <= count; index++) {
            setPageList(prevState => [...prevState, index])
        }
    }, [count])

    useEffect(() => {
        onChange(currentPage)
    }, [currentPage])

    const handlePrev = () => {
        if (currentPage > 1) {
            setCurrentPage(prevState => prevState - 1)
        }
    }

    const handleNext = () => {
        if (currentPage < count) {
            setCurrentPage(prevState => prevState + 1)
        }
    }

    return (
        <Container theme={theme}>
            <ActionButton theme={theme} onClick={handlePrev}>
                <BsArrowLeftShort />
            </ActionButton>
            {count <= 5 && (
                <>
                    {pageList.map((page, index) => (
                        <PaginationItem
                            theme={theme}
                            key={index}
                            isActive={page === currentPage}
                            onClick={() => setCurrentPage(page)}
                        >
                            {page}
                        </PaginationItem>
                    ))}
                </>
            )}
            {count > 5 && (
                <>
                    <PaginationItem
                        theme={theme}
                        isActive={pageList[0] === currentPage}
                        onClick={() => setCurrentPage(pageList[0])}
                    >
                        {pageList[0]}
                    </PaginationItem>
                    {currentPage >= 5 ? (
                        <>
                            <PaginationItem
                                theme={theme}
                                isActive={false}
                                disabled={true}
                            >
                                ...
                            </PaginationItem>
                            {currentPage <= count - 3 && (
                                <>
                                    <PaginationItem
                                        theme={theme}
                                        isActive={
                                            pageList[currentPage - 2] ===
                                            currentPage
                                        }
                                        onClick={() =>
                                            setCurrentPage(
                                                pageList[currentPage - 2]
                                            )
                                        }
                                    >
                                        {pageList[currentPage - 2]}
                                    </PaginationItem>
                                    <PaginationItem
                                        theme={theme}
                                        isActive={
                                            pageList[currentPage - 1] ===
                                            currentPage
                                        }
                                        onClick={() =>
                                            setCurrentPage(
                                                pageList[currentPage - 1]
                                            )
                                        }
                                    >
                                        {pageList[currentPage - 1]}
                                    </PaginationItem>
                                    <PaginationItem
                                        theme={theme}
                                        isActive={
                                            pageList[currentPage] ===
                                            currentPage
                                        }
                                        onClick={() =>
                                            setCurrentPage(
                                                pageList[currentPage]
                                            )
                                        }
                                    >
                                        {pageList[currentPage]}
                                    </PaginationItem>
                                </>
                            )}
                        </>
                    ) : (
                        <>
                            {pageList
                                .filter(element => element >= 2 && element <= 5)
                                .map((page, index) => (
                                    <PaginationItem
                                        theme={theme}
                                        key={index}
                                        isActive={page === currentPage}
                                        onClick={() => setCurrentPage(page)}
                                    >
                                        {page}
                                    </PaginationItem>
                                ))}
                        </>
                    )}
                    {currentPage <= count - 3 ? (
                        <>
                            <PaginationItem
                                theme={theme}
                                isActive={false}
                                disabled={true}
                            >
                                ...
                            </PaginationItem>

                            <PaginationItem
                                theme={theme}
                                isActive={pageList[count - 1] === currentPage}
                                onClick={() =>
                                    setCurrentPage(pageList[count - 1])
                                }
                            >
                                {pageList[count - 1]}
                            </PaginationItem>
                        </>
                    ) : (
                        <>
                            {pageList
                                .filter(element => element >= count - 4)
                                .map((page, index) => (
                                    <PaginationItem
                                        theme={theme}
                                        key={index}
                                        isActive={page === currentPage}
                                        onClick={() => setCurrentPage(page)}
                                    >
                                        {page}
                                    </PaginationItem>
                                ))}
                        </>
                    )}
                </>
            )}

            <ActionButton theme={theme} onClick={handleNext}>
                <BsArrowRightShort />
            </ActionButton>
        </Container>
    )
}

export default PaginationComponent
