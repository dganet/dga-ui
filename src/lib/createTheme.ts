import { defaultTheme } from '../styles/defaultTheme'

export const createTheme = (theme: typeof defaultTheme) => {
    return {
        ...defaultTheme,
        colors: {
            ...defaultTheme.colors,
            ...theme.colors
        },

        name: theme.name || defaultTheme.name
    } as typeof defaultTheme
}
