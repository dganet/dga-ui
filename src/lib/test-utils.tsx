import { render, RenderOptions } from '@testing-library/react'
import React, { ReactElement, FC } from 'react'
import { defaultTheme } from '../styles/defaultTheme'
import { CustomThemeProvider } from '../hooks/useTheme/ThemeContext'

const Providers: FC = ({ children }) => {
    return (
        <CustomThemeProvider theme={defaultTheme}>
            {children}
        </CustomThemeProvider>
    )
}

/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: Providers, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
