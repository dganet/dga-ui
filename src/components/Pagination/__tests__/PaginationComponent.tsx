import { render, screen } from '../../../lib/test-utils'

import PaginationComponent from '../PaginationComponent'

describe('<PaginationComponent />', () => {
  it('should render the component', () => {
    const { container } = render(<PaginationComponent />)

    expect(screen.getByRole('heading', { name: /Pagination/i })).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
