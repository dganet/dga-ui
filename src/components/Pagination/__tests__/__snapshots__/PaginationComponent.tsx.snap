// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PaginationComponent /> should render the component 1`] = `
<div
  class="sc-bdfBQB"
>
  <h1>
    Pagination - 
    new component
  </h1>
</div>
`;
